### ISSUE TITLE
[red-knot] instance attributes bound in nested scopes are reported as unresolved

### ISSUE THREAD

**Issue by mtshiba** [astral-sh/ruff#17008]
### Summary

From https://github.com/astral-sh/ruff/pull/16852#issuecomment-2756606180

```python
class C:
    def __init__(self):
        def closure():
            self.a: str = 1
        closure()
        [... for self.b in range(1)]
        class D:
            self.c = 1

# should be OK
C().a
C().b
C().c
```

```sh
error: lint:unresolved-attribute
  --> test.py:10:7
   |
 8 |             self.c = 1
 9 |
10 | print(C().a)
   |       ^^^^^ Type `C` has no attribute `a`
11 | print(C().b)
12 | print(C().c)
   |

error: lint:unresolved-attribute
  --> test.py:11:7
   |
10 | print(C().a)
11 | print(C().b)
   |       ^^^^^ Type `C` has no attribute `b`
12 | print(C().c)
   |

error: lint:unresolved-attribute
  --> test.py:12:7
   |
10 | print(C().a)
11 | print(C().b)
12 | print(C().c)
   |       ^^^^^ Type `C` has no attribute `c`
   |

Found 3 diagnostics
```

### Version

red-knot 0.0.0 (640d821108b6c4ed223b4b8ca1153facb558f870)
